### PR TITLE
WIP: Add warnMutations middleware

### PR DESCRIPTION
**Pull request by leoasis**
Aims to fix https://github.com/gaearon/redux/issues/135

I've implemented the mutation warnings as a middleware, which expects to receive plain actions (should be the last in the chain or at least should be after the ones that transform the actions such as the `thunkMiddleware`).

Still need to provide support for stuff like ImmutableJS but should be straight forward. Also, I'm leaving it like this so that we can discuss implementation.

We can also discuss what's the best way to handle adding this middleware by default on dev, but not in prod (and avoid including the dependencies as well).

### CODE REVIEW

**Comment by goatslacker** [comment]
Copying deep on every dispatch is probably not feasible unless you're building apps with small state.

I can't think of another way to do this off the top of my head though.

**Comment by leoasis** [comment]
This is meant to be used in development only.

**Comment by gaearon** [comment]
This assumes `dispatch` is sync so it needs to be the last in chain right?

**Comment by leoasis** [comment]
@gaearon yes: 
> I've implemented the mutation warnings as a middleware, which expects to receive plain actions (should be the last in the chain or at least should be after the ones that transform the actions such as the thunkMiddleware).

File: src/middleware/warnMutations.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import isEqual from 'lodash/lang/isEqual';
+import any from 'lodash/collection/any';
+import cloneDeep from 'lodash/lang/cloneDeep';
```
Comment (taylorhakes) [inline src/middleware/warnMutations.js]: A recent pull request removed lodash dependency. I think it would be better to avoid using lodash

File: src/middleware/warnMutations.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      ].join(''));
+    }
+
+    action = next(action);
+
+    lastStateRef = getState();
+    lastState = copyState(lastStateRef);
+
+    if (wasMutated(stateRef, state, lastStateRef, lastState)) {
+      console.warn(
```
Comment (devknoll) [inline src/middleware/warnMutations.js]: We're already using `invariant`, do we want to add `warning` as well?

File: src/middleware/warnMutations.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  let lastStateRef = getState();
+  let lastState = copyState(lastStateRef);
+
+  return (next) => (action) => {
+    const stateRef = getState();
+    const state = copyState(stateRef);
+
+    if (wasMutated(lastStateRef, lastState, stateRef, state)) {
+      console.warn([
+        'A state mutation was detected between dispatches.' +
```
Comment (taylorhakes) [inline src/middleware/warnMutations.js]: mixed of comma and plus operators

File: src/middleware/warnMutations.js (hunk 2 above)
Comment (taylorhakes) [inline src/middleware/warnMutations.js]: I would prefer to use invariant. If I am using this middleware, I would prefer to get a hard error. This middleware isn't used by default, so it won't cause issues in prod.

File: src/middleware/warnMutations.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    lastStateRef = getState();
+    lastState = copyState(lastStateRef);
+
+    if (wasMutated(stateRef, state, lastStateRef, lastState)) {
+      console.warn(
+        'A state mutation was detected inside a dispatch.',
+        ` Take a look at the store(s) handling the action`,
+        action,
+        '(https://github.com/gaearon/redux#my-views-arent-updating)'
```
Comment (ooflorent) [inline src/middleware/warnMutations.js]: Please change this to direct links to the doc file

File: src/middleware/warnMutations.js (hunk 2 above)
Comment (leoasis) [inline src/middleware/warnMutations.js]: makes sense, will use that then

File: src/middleware/warnMutations.js (hunk 4 above)
Comment (leoasis) [inline src/middleware/warnMutations.js]: what do you mean? It's already showing as a link

**Comment by leoasis** [comment]
Pushed some updates: Avoiding using lodash and checking for immutable types.

I'm not sure if it's a good thing to avoid using libraries that have been battle tested like lodash in spite of library size. We'll be maintaining and having to support edge cases for deep cloning or whatever we're replacing from lodash. Still, perhaps that's enough for our needs in this case, so if anyone sees something odd here or anything I'm missing let me know! I'll try to add more test cases to see I'm not missing any basic scenario

**Comment by gaearon** [comment]
>We'll be maintaining and having to support edge cases for deep cloning or whatever we're replacing from lodash. 

I only think it makes sense to avoid lodash for the core stuff. Plugins (which should be in separate repos anyway) can use lodash, especially if they're dev-only.

**Comment by acdlite** [comment]
If/when you do use lodash, you should also require the specific module you want to use.

```js
// This
import isEqual from "lodash.isequal"

// instead of this
import isEqual from "lodash/lang/isEqual"
```

**Comment by leoasis** [comment]
> If/when you do use lodash, you should also require the specific module you want to use.
@acdlite what's the reason for that? the latter option will still make the bundle require only the needed modules and no more

**Comment by gaearon** [comment]
>the latter option will still make the bundle require only the needed modules and no more

This is correct AFAIK.

**Comment by acdlite** [comment]
True, but the first way it makes it less likely another dev will see a lodash dependecy and be tempted to import the entire library elsewhere.

**Comment by johanneslumpe** [comment]
To me personally there is no difference between the two versions, except that the latter is more descriptive. If at all, the second one would be more logical to me, as people see that you are specifically importing a subset/function directly, instead of a prop from the whole lib.

**Comment by acdlite** [comment]
Think about what the corresponding `package.json` files look like. The first has `lodash.isequal` and the second has `lodash`. Another developer may see a dependency on lodash and assume its safe to import the whole library, whereas a dependency on `lodash.isequal` forces restraint. Clearly this is a matter of preference, though.

**Comment by leoasis** [comment]
I agree it's mostly a matter of preference. The only concern I have is that
it is very likely that the end user will already be using lodash (compared
to using the separate function modules), and if we depend on just the
function module we will be making the final user bundle even larger since
it will contain duplicated stuff that webpack or any other bundler won't be
able to realize it's the same.

But again since this particular thing is just for development, that
argument perhaps it's not that important.

**Comment by acdlite** [comment]
Huh, I never import the entire lodash module ¯\_(ツ)_/¯

**Comment by leoasis** [comment]
Not meant import, but have lodash as a dependency in package.json

**Comment by leoasis** [comment]
In any case, since the argument I mentioned is not strong in this case
since this middleware is meant to be used in development only, let's go
with the function module dependency

**Comment by leoasis** [comment]
Ok I think this is now mostly complete. We now need to decide if we'll leave it here, move it to a different repo, and if leaving it here if we should automatically add it to the default redux middleware if we detect we're in the development env (assuming `process.env.NODE_ENV !== 'production'`?).

**Comment by gaearon** [comment]
Please can you move this to a separate repo and try to get it working with the `breaking-changes-1.0` branch? I'm closing for now, as it doesn't belong in the core.

**Comment by leoasis** [comment]
@gaearon will do! Any particular name for this one you would prefer? `redux-warn-mutations`? `redux-dev-warn-mutations` (to make it clear that it's just for dev)?

**Comment by 0xR** [comment]
@leoasis's library is here in case you are looking for it in this thread:
https://github.com/leoasis/redux-immutable-state-invariant

**Comment by leoasis** [comment]
@0xR thanks for adding this here
